File: test/codegen/py/math.test.py
#!/usr/bin/python3

import unittest

from compiled.math import avg, compose, mean, msort, neg, qsort, total, tail


class TestMath(unittest.TestCase):

    def test_avg(self):
        self.assertEqual(avg(6)(10), 8)

    def test_compose(self):
        self.assertEqual(compose(lambda x: x + 1)(lambda y: y * 2)(3), 7)

    def test_mean(self):
        self.assertEqual(mean([5, 10, 15]), 10)

    def test_quadruple(self):
        self.assertEqual(msort([1,4,2,5,7,8,3,6]), [1,2,3,4,5,6,7,8])

    def test_neg(self):
        self.assertEqual(neg(7), -7)

    def test_quadruple(self):
        self.assertEqual(qsort([1,4,2,5,7,8,3,6]), [1,2,3,4,5,6,7,8])

    def test_total(self):
        self.assertEqual(total([1,2,3,4,5]), 15)

    def test_tail(self):
        self.assertEqual(tail([1,2,3]), [2,3])

if __name__ == '__main__':
    unittest.main()

File: test/codegen/lua/math.test.lua
#!/usr/bin/env lua

require "test.codegen.lua.compiled.math"
require "test.codegen.lua.unittest"

assertEqual("avg", avg(6)(10), 8)
assertEqual("compose", compose(_plus_(1))(_mult_(2))(3), 7)
assertEqual("mean", mean({head = 5, tail = { head = 10, tail = { head = 15, tail = {}}}}), 10)
local list = { head = 1, tail = { head = 4 , tail = { head = 2, tail = {}}}}
local sortedList = { head = 1, tail = { head = 2 , tail = { head = 4, tail = {}}}}
assertEqual("msort", msort(list), sortedList)
assertEqual("qsort", qsort(list), sortedList)
assertEqual("neg", neg(7), -7)
assertEqual("total", total({ head = 8, tail = { head = 4 , tail = { head = 7, tail = {}}}}), 19)
assertEqual("tail", tail({ head = 1, tail = { head = 2 , tail = { head = 3, tail = {}}}}), { head = 2 , tail = { head = 3, tail = {}}})

print("PASSED", "math.test.lua")
